### PR TITLE
fix discount factors for historical dates

### PR DESCRIPTION
**Pull request by benhamilton-brown**

### CODE REVIEW

**Comment by yukiiwashita** [comment]
This change to DiscountFactors involves a potential risk - we never get a value estimated from a curve for negative year fractions. If you really need discountFactor=1 for negative year fractions, is this something we can control in the underlying curve, e.g., 0 for negative x values for a zero rate curve, etc.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRateDiscountFactors.java [hunk 1]
```diff
@@ -168,11 +170,17 @@ public double relativeYearFraction(LocalDate date) {
   @Override
   public double discountFactor(double yearFraction) {
     // convert zero rate to discount factor
+    if (yearFraction <= EFFECTIVE_ZERO) {
```
Comment (yukiiwashita) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRateDiscountFactors.java:173]: If `discountFactor` is modified, `zeroRate` needs to be also modified for consistency. By the current change, we have discountFactor = 1, but nonzero yearFraction*zeroRate, for negative year fractions.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRatePeriodicDiscountFactors.java [hunk 2]
```diff
@@ -194,14 +200,18 @@ public double discountFactorTimeDerivative(double yearFraction) {
 
   @Override
   public double zeroRate(double yearFraction) {
-    double ratePeriod = curve.yValue(yearFraction);
+    double yearFractionMod = Math.max(EFFECTIVE_ZERO, yearFraction);
+    double ratePeriod = curve.yValue(yearFractionMod);
```
Comment (yukiiwashita) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRatePeriodicDiscountFactors.java:204]: Is this consitent with `discountFactor` and `zeroRatePointSensitivity`?

**Review by yukiiwashita — CHANGES_REQUESTED** [review]

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRatePeriodicDiscountFactors.java (hunk 2 above)
Comment (benhamilton-brown) [reply modules/pricer/src/main/java/com/opengamma/strata/pricer/ZeroRatePeriodicDiscountFactors.java]: I have replaced the above and done the same in ZeroRateDiscountFactors.java to ensure we always return 0 if year faction is not greater than 0.

**Review by yukiiwashita — APPROVED** [review]

**Pull request merged**